R/utilities-sensitivity-analysis.R

```diff
@@ -1,342 +1,343 @@
+  allResultsFileNames <- generateResultFileNames(numberOfCores = numberOfCores, folderName = resultsFileFolder, fileName = resultsFileName)
+
+
+  Rmpi::mpi.bcast.Robj2slave(obj = allResultsFileNames)
+
+  # Load simulation on each core
+  Rmpi::mpi.bcast.cmd(sim <- loadSimulation(simFilePath))
+
+  # Update simulation with individual parameters
+  Rmpi::mpi.bcast.cmd(updateSimulationIndividualParameters(simulation = sim, individualParameters))
+
+
+
+  Rmpi::mpi.remote.exec(analyzeCoreSensitivity(
+    simulation = sim,
+    parametersToPerturb = listSplitParameters[[mpi.comm.rank()]],
+    totalSensitivityThreshold = 1,
+    resultsFilePath = allResultsFileNames[mpi.comm.rank()],
+    numberOfCoresToUse = 1 # Number of local cores, set to 1 when parallelizing.
+  ))
+  Rmpi::mpi.close.Rslaves()
+
+  allSAResults <- importSensitivityAnalysisResultsFromCSV(simulation = loadSimulation(simFilePath), filePaths = allResultsFileNames)
+  combinedFilePath <- file.path(resultsFileFolder, paste0(resultsFileName, ".csv"))
+  exportSensitivityAnalysisResultsToCSV(results = allSAResults, filePath = combinedFilePath)
+  file.remove(allResultsFileNames)
+  return(combinedFilePath)
+}
+
+#' @title analyzeCoreSensitivity
+#' @description Run a sensitivity analysis for a single individual, perturbing only the set of parameters parametersToPerturb
+#' @param simulation simulation class object
+#' @param parametersToPerturb paths of parameters to be analyzed
+#' @param totalSensitivityThreshold numeric value between 0 and 1.
+#' Close to 0, only the most sensitive output paths are returned.
+#' Close to 1, almost all the output paths are returned.
+#' @param resultsFilePath Path to file storing results of sensitivity analysis
+#' @param numberOfCoresToUse Number of cores to use on local node.  This parameter should be should be set to 1 when parallelizing over many nodes.
+#' @return Save sensitivity analysis results as CSV in path given by resultsFilePath.
+#' @import ospsuite
+#' @export
+analyzeCoreSensitivity <- function(simulation,
+                                   parametersToPerturb = NULL,
+                                   totalSensitivityThreshold = 1,
+                                   resultsFilePath = paste0(getwd(), "sensitivityAnalysisResults.csv"),
+                                   numberOfCoresToUse = NULL) {
+  sensitivityAnalysis <- SensitivityAnalysis$new(simulation = simulation)
+  sensitivityAnalysis$addParameterPaths(parametersToPerturb)
+
+
+
+  if (is.null(numberOfCoresToUse)) {
+    sensitivityAnalysisRunOptions <- SensitivityAnalysisRunOptions$new(showProgress = FALSE)
+  }
+  else {
+    sensitivityAnalysisRunOptions <- SensitivityAnalysisRunOptions$new(
+      showProgress = FALSE,
+      numberOfCoresToUse = numberOfCoresToUse
+      #The numberOfCoresToUse input in the SensitivityAnalysisRunOptions initializer should not be set to NULL.
```

Comment:
Interesting. That looks like a bug to me don't you think?

Reply:
Seems so